downloader/download.go

```diff
@@ -2,13 +2,268 @@ package downloader
 
 import (
 	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
 
 	"github.com/src-d/gitcollector/library"
+	"github.com/src-d/go-borges"
+	"github.com/src-d/go-borges/siva"
+	"gopkg.in/src-d/go-billy.v4"
+	"gopkg.in/src-d/go-billy.v4/util"
+	"gopkg.in/src-d/go-errors.v1"
+	"gopkg.in/src-d/go-git.v4"
+)
+
+var (
+	// ErrNotDownloadJob is returned when a not download job is found.
+	ErrNotDownloadJob = errors.NewKind("not download job")
+
+	// ErrRepoAlreadyExists is returned if there is an attempt to
+	// retrieve an already downloaded git repository.
+	ErrRepoAlreadyExists = errors.NewKind("%s already downloaded")
+
+	// ErrNeedSivaLibrary is returned when a siva borges.Library is needed.
+	ErrNeedSivaLibrary = errors.NewKind("a siva library is needed")
 )
 
 // Download is a library.JobFn function to download a git repository and store
 // it in a borges.Library.
-func Download(_ context.Context, _ *library.Job) error {
-	// TODO: implement repsitories download
+func Download(_ context.Context, job *library.Job) error {
+	if len(job.Endpoints) == 0 || job.Lib == nil || job.TempFS == nil {
+		return ErrNotDownloadJob.New()
+	}
+
+	lib, ok := (job.Lib).(*siva.Library)
+	if !ok {
+		return ErrNeedSivaLibrary.New()
+	}
+
+	return downloadRepo(lib, job.TempFS, job.Endpoints[0])
```

Comment:
We should open an issue to use context for cancellation / timeout in downloads.

Reply:
Issue opened at #12